File: android/core/export/src/main/kotlin/jp/rimtty/codematch/core/export/ReportMailContent.kt
package jp.rimtty.codematch.core.export

import jp.rimtty.codematch.core.model.AppLanguage
import jp.rimtty.codematch.core.model.Destination
import jp.rimtty.codematch.core.model.MatchSession
import java.time.ZoneId

/** Subject and body of the e-mail a report is shared with. */
data class ReportMail(
    val subject: String,
    val body: String,
)

/**
 * Builds the pre-filled e-mail for sharing a session's PDF.
 *
 * The recipient is fixed for now: the report is only ever sent to the
 * operator's own mailbox. The body repeats the report header (session,
 * destination, box and row counts) so the mail is readable without opening
 * the attachment. Sending itself stays with the mail app the operator uses.
 *
 * This mirrors Swift's `ReportMailContent`.
 */
object ReportMailContent {
    /** The one address every report mail is addressed to. */
    const val RECIPIENT: String = "ttyrim@gmail.com"

    val recipients: Array<String>
        get() = arrayOf(RECIPIENT)

    fun build(
        session: MatchSession,
        kind: HistoryReportKind,
        fileName: String,
        language: AppLanguage = AppLanguage.JAPANESE,
        zoneId: ZoneId = ZoneId.systemDefault(),
    ): ReportMail {
        val labels = HistoryExportTextFormatter.labels(language)
        val title = when (kind) {
            HistoryReportKind.MATCH_HISTORY -> labels.reportTitle
            HistoryReportKind.INSPECTION -> labels.inspectionTitle
        }
        val destination = session.resolvedDestination()
        val sessionLabel = session.displayName.ifBlank {
            HistoryExportTextFormatter.dateTime(session.startedAt, language, zoneId)
        }
        val subject = buildString {
            append("[CodeMatch] ").append(title).append(' ').append(sessionLabel)
            if (destination != null) append(" - ").append(labels.destinationName(destination))
        }

        val lines = mutableListOf<String>()
        lines += labels.mailGreeting
        lines += when (kind) {
            HistoryReportKind.MATCH_HISTORY -> labels.mailIntroHistory
            HistoryReportKind.INSPECTION -> labels.mailIntroInspection
        }
        lines += ""
        lines += labels.mailSessionHeading
        if (session.displayName.isNotEmpty()) {
            lines += "${labels.sessionName}: ${session.displayName}"
        }
        lines += "${labels.start}: ${HistoryExportTextFormatter.dateTime(session.startedAt, language, zoneId)}"
        val endedAt = session.endedAt
        lines += if (endedAt != null) {
            "${labels.end}: ${HistoryExportTextFormatter.dateTime(endedAt, language, zoneId)}"
        } else {
            "${labels.status}: ${labels.inProgress}"
        }
        if (destination != null) {
            lines += "${labels.destination}: ${labels.destinationName(destination)}"
        }
        lines += "${labels.inspectionBoxCount}: ${HistoryExportTextFormatter.boxCount(session.matchedCount, language)}"
        lines += countLines(session, kind, destination, labels, language)
        lines += ""
        lines += labels.mailAttachmentHeading
        lines += fileName
        lines += ""
        lines += labels.mailFooter

        return ReportMail(subject = subject, body = lines.joinToString("\n"))
    }

    /** The same count lines the report's own header prints. */
    private fun countLines(
        session: MatchSession,
        kind: HistoryReportKind,
        destination: Destination?,
        labels: HistoryExportLabels,
        language: AppLanguage,
    ): List<String> = when (kind) {
        HistoryReportKind.INSPECTION -> {
            val report = InspectionReportContent.build(session)
            val label = when (report.layout) {
                InspectionLayout.SAWAI -> labels.inspectionPartCountBySuffix
                InspectionLayout.MOLTEN -> labels.deliveryNumberCount
                InspectionLayout.DENSO -> labels.inspectionPartCount
            }
            listOf("$label: ${HistoryExportTextFormatter.integer(report.rowCount, language)}")
        }

        HistoryReportKind.MATCH_HISTORY -> buildList {
            add("${labels.inspectionPartCount}: ${HistoryExportTextFormatter.integer(session.groupedEntries.size, language)}")
            if (destination == Destination.MOLTEN) {
                val count = session.entries.moltenDeliveryGroups().size
                add("${labels.deliveryNumberCount}: ${HistoryExportTextFormatter.integer(count, language)}")
            }
        }
    }
}

File: ios/CodeMatch/Services/ReportMailContent.swift
import Foundation

/// セッション詳細から書き出す2種類のPDF。
enum ReportKind {
    case matchHistory
    case inspection
}

/// レポートPDFをメールで共有するときの宛先・件名・本文。
///
/// 宛先は当面固定で、レポートは操作者自身の受信箱にだけ送る。本文にはレポートの見出し
/// （セッション、仕向地、箱数・行数）を繰り返し、添付を開かなくても内容が分かるようにする。
/// 送信そのものは操作者が使うメールアプリに任せる。Android の `ReportMailContent` と同じ規則。
struct ReportMailContent: Equatable {
    /// すべてのレポートメールの宛先。
    static let recipients = ["takemoto1075@icloud.com"]

    let subject: String
    let body: String

    static func make(session: MatchSession, kind: ReportKind, fileName: String, locale: Locale) -> ReportMailContent {
        let appLanguage = AppLanguage(locale)
        let title: String
        switch kind {
        case .matchHistory: title = AppLocalization.string("照合履歴レポート")
        case .inspection: title = AppLocalization.string("検品レポート")
        }
        let destination = session.resolvedDestination
        // 「検品レポート 2026/09/07 23:17 - 澤井製作所」。セッション名ではなく開始日時で並ぶようにする。
        var subject = "\(title) \(appLanguage.formatDateTime(session.startedAt))"
        if let destination {
            subject += " - \(destination.displayName)"
        }

        var lines: [String] = []
        lines.append(AppLocalization.string("お疲れさまです。"))
        switch kind {
        case .matchHistory: lines.append(AppLocalization.string("CodeMatch の照合履歴レポートをお送りします。"))
        case .inspection: lines.append(AppLocalization.string("CodeMatch の検品レポートをお送りします。"))
        }
        lines.append("")
        lines.append(AppLocalization.string("■ セッション"))
        if !session.displayName.isEmpty {
            lines.append("\(AppLocalization.string("セッション名")): \(session.displayName)")
        }
        lines.append("\(AppLocalization.string("開始")): \(appLanguage.formatDateTime(session.startedAt))")
        if let endedAt = session.endedAt {
            lines.append("\(AppLocalization.string("終了")): \(appLanguage.formatDateTime(endedAt))")
        } else {
            lines.append("\(AppLocalization.string("状態")): \(AppLocalization.string("照合中"))")
        }
        if let destination {
            lines.append(AppLocalization.string("仕向地: \(destination.displayName)"))
        }
        lines.append(AppLocalization.string("検査箱数: \(session.matchedCount)箱"))
        lines.append(contentsOf: countLines(session: session, kind: kind, destination: destination))
        lines.append("")
        lines.append(AppLocalization.string("■ 添付"))
        lines.append(fileName)

        return ReportMailContent(subject: subject, body: lines.joined(separator: "\n"))
    }

    /// レポート自身の見出しと同じ件数の行。
    private static func countLines(session: MatchSession, kind: ReportKind, destination: Destination?) -> [String] {
        switch kind {
        case .inspection:
            let report = InspectionReport.make(session: session)
            switch report.layout {
            case .sawai: return [AppLocalization.string("品番数（枝番別）: \(report.rowCount)")]
            case .molten: return [AppLocalization.string("納品番号数: \(report.rowCount)")]
            case .denso: return [AppLocalization.string("品番数: \(report.rowCount)")]
            }
        case .matchHistory:
            var lines = [AppLocalization.string("品番数: \(session.groupedEntries.count)")]
            if destination == .molten {
                lines.append(AppLocalization.string("納品番号数: \(session.deliveryNumberCount)"))
            }
            return lines
        }
    }
}
